backends/p4tools/common/lib/formulae.cpp

```diff
@@ -0,0 +1,130 @@
+    return compare(node.get(), other.node.get()) < 0;
+}
+
+bool StateVariable::operator==(const StateVariable& other) const {
+    // Delegate to IR's notion of equality.
+    return *node == *other.node;
+}
+
+int StateVariable::compare(const IR::Expression* e1, const IR::Expression* e2) {
+    if (const auto* m1 = e1->to<IR::Member>()) {
+        return compare(m1, e2);
+    }
+    if (const auto* p1 = e1->to<IR::PathExpression>()) {
+        return compare(p1, e2);
+    }
+    BUG("Not a valid StateVariable: %1%", e1);
+}
+
+int StateVariable::compare(const IR::Member* m1, const IR::Expression* e2) {
+    if (const auto* m2 = e2->to<IR::Member>()) {
+        return compare(m1, m2);
+    }
+    if (e2->is<IR::PathExpression>()) {
+        return 1;
+    }
+    BUG("Not a valid StateVariable: %1%", e2);
+}
+
+int StateVariable::compare(const IR::Member* m1, const IR::Member* m2) {
+    auto result = compare(m1->expr, m2->expr);
+    if (result != 0) {
+        return result;
+    }
+    if (m1->member.name < m2->member.name) {
+        return -1;
+    }
+    if (m1->member.name > m2->member.name) {
+        return 1;
+    }
+    return 0;
+}
+
+int StateVariable::compare(const IR::PathExpression* p1, const IR::Expression* e2) {
+    if (const auto* p2 = e2->to<IR::PathExpression>()) {
+        return compare(p1, p2);
+    }
+    if (e2->is<IR::Member>()) {
+        return -1;
+    }
+    BUG("Not a valid StateVariable: %1%", e2);
+}
+
+int StateVariable::compare(const IR::PathExpression* p1, const IR::PathExpression* p2) {
+    if (p1->path->name.name < p2->path->name.name) {
+        return -1;
+    }
+    if (p1->path->name.name > p2->path->name.name) {
+        return 1;
+    }
+    return 0;
+}
+
+/* =============================================================================================
+ *  ConcolicVariable
+ * ============================================================================================= */
+
+ConcolicVariable::ConcolicVariable(const IR::Expression* expr, const IR::MethodCallExpression* call,
+                                   uint64_t instanceId)
+    : AbstractRepCheckedNode(checkedTo<IR::Member>(expr), "concolic variable"),
+      call(call),
+      instanceId(instanceId) {}
+
+bool ConcolicVariable::repOk(const IR::Expression* expr) {
+    // Only members can be state variables.
+    const auto* member = expr->to<IR::Member>();
+    if (member == nullptr) {
+        return false;
+    }
+
+    // A member is a state variable if it is qualified by a PathExpression or if its qualifier is a
+    // state variable.
+    return member->expr->is<IR::PathExpression>() || repOk(member->expr);
+}
+
+const IR::MethodCallExpression* ConcolicVariable::getCall() const { return call; }
+
+uint64_t ConcolicVariable::getInstanceId() const { return instanceId; }
+
+bool ConcolicVariable::operator<(const ConcolicVariable& other) const {
+    if (node->toString() < other.node->toString()) {
```

Comment:
in general tostring does not serialize completely an expression, this may be a bug waiting to strike.

Reply:
I believe this code was actually superseded by a definition in `common/testgen.def`. Thanks for catching this.